app/scripts/modules/core/src/application/application.model.ts

```diff
@@ -118,7 +122,7 @@ export class Application {
     };
   }
 
-  private addDataSource(config: IDataSourceConfig) {
+  private addDataSource(config: IDataSourceConfig<any>) {
```

Comment:
If there are a finite number of data sources, my suggestion would be to either create an enum with those possible types to make this function more strict:
```
interface DataSources = IServerGroup | ILoadBalancer | IAnotherDataSource
private addDataSource(config: IDataSourceConfig<DataSources> { ... }
```
Or to introduce generics to this function as well. This one seems trickier based on when the function is invoked.
```
private addDataSource<T>(config: IDataSourceConfig<T>) { ... }
```

Reply:
You're right that using generics here would be the most accurate typing, and I'll do a quick update to reflect that. Sadly, though, there really isn't any way to type this strictly enough that we get meaningful type info beyond `any` because:

- There is not a finite list of available data sources — some of them are added in provider modules outside core, others are added in repos outside `deck` like `deck-kayenta`, and still more are added in custom builds (like our internal build at Netflix)
- When the application model calls this method, it's doing it by looping over the list of all available data sources, which means the only type we can use to appropriately describe the contents is `any`

The 'big list of arbitrary data polling wrappers' architecture that currently exists doesn't lend itself to very useful static typing, but it does make it possible to coordinate polling across a lot of different endpoints beyond what was originally written into core. Not saying it's the perfect tradeoff today in 2019, but it has helped us do interesting things in the past.

Reply:
Got it, thanks for the additional context here.